ikarus/python/finiteelements/material.hh

```diff
@@ -17,131 +17,187 @@
-      Dune::Python::insertClass<Ikarus::Materialname<args>>(                                                 \
-          m, #Materialname, Dune::Python::GenerateTypeName("Ikarus::" + std::string(#Materialname<##args>)), \
-          includes##Materialname)                                                                            \
-          .first;                                                                                            \
-  cls##Materialname.def(pybind11::init([](double emod, double nu) {                                          \
-                          auto matParameter =                                                                \
-                              Ikarus::toLamesFirstParameterAndShearModulus({.emodul = emod, .nu = nu});      \
-                          return new Materialname(matParameter);                                             \
-                        }),                                                                                  \
-                        "Material constructor that takes Young's modulus E and Poisson's ratio nu"           \
-                        "E"_a,                                                                               \
-                        "nu"_a);                                                                             \
-  MAKE_MaterialFunction(Materialname<##args>, storedEnergy);                                                 \
-  MAKE_MaterialFunction(Materialname<##args>, stresses);                                                     \
-  MAKE_MaterialFunction(Materialname<##args>, tangentModuli);
+  Ikarus::LamesFirstParameterAndShearModulus extractMaterialParameters(const pybind11::kwargs& kwargs) {
+    // clang-format off
+    static const std::map<std::pair<std::string, std::string>, std::function<LamesFirstParameterAndShearModulus(const pybind11::kwargs&)>> conversionMap = {
+      {{"E", "nu"}, [](const auto& kw){ return convertMaterialParameters<YoungsModulusAndPoissonsRatio>(kw, "E", "nu"); }},
+      {{"E", "mu"}, [](const auto& kw){ return convertMaterialParameters<YoungsModulusAndShearModulus>(kw, "E", "mu"); }},
+      {{"E", "K"}, [](const auto& kw){ return convertMaterialParameters<YoungsModulusAndBulkModulus>(kw, "E", "K"); }},
+      {{"E", "Lambda"}, [](const auto& kw){ return convertMaterialParameters<YoungsModulusAndLamesFirstParameter>(kw, "E", "Lambda"); }},
+      {{"K", "Lambda"}, [](const auto& kw){ return convertMaterialParameters<BulkModulusAndLamesFirstParameter>(kw, "K", "Lambda"); }},
+      {{"Lambda", "mu"}, [](const auto& kw){ return LamesFirstParameterAndShearModulus{kw["Lambda"].template cast<double>(), kw["mu"].template cast<double>()}; }}
+    };
+    // clang-format off
 
-namespace Ikarus::Python {
+  if (kwargs.size() != 2)
+    throw std::runtime_error("The number of material parameters passed to the material should be 2");
+
+  for (const auto& conversion : conversionMap) {
```

Comment:
Couldn't this be "just"
```cpp
return conversionMap({kwargs[pybind11::cast(0)].cast<std::string>(),kwargs[pybind11::cast(1)].cast<std::string>()})(kwargs); 
```
? 

I think you can add a try catch with `throw error_already_set` and the rethrow your nice error message below.

Reply:
I couldn't get this to work, I am not sure if we can index the dict with numbers, also then the user has to pass in the material parameters in a certain order

Reply:
If you really stop resolving the conversatrions them I would see your comments^^

Reply:
> If you really stop resolving the conversatrions them I would see your comments^^

Sorry! 

> Regarding the indexing with integer this should work which I looked up. What problems appeared?
I get
```bash
File "/workspaces/ikarus/ikarus/python/test/testmaterials.py", line 167, in checkWithStrain
  svk = iks.materials.StVenantKirchhoff(E=1000, nu=0.3)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
KeyError: 0
```

Reply:
> I think the old version also allows only a single order, since the key in the map is the pair in this specific order.

At least this seems to work 
```py
    svk = iks.materials.StVenantKirchhoff(nu= 0.3, E=1000)
```

Reply:
Ok my browser crashed twice removing my comment draft.

Reply:
So you are right that the order is not important and also the integer indexing is not working like that.

Reply:
Can you still pleas refactor like this. I think the `std::pair` is not needed sicne both types are the same.
Furthermore, structured bindings and variable names would also make it clearer.

```cpp
    std::map < std::array<std::string, 2>, std::function<...> dict;

    for (const auto [materialParameters, parameterTransformer] : dict) {
        const auto [firstPar, secondPar] = materialParameters;
        if (kwargs.contains(firstPar) and kwargs.contains(secondPar))
            return parameterTransformer(kwargs)
    }
   
    ```